REST: Document mutable inputs on Create() and fix tests

thockin opened the pull request:
If one doesn't DeepCopy() on the way into Create, we can end up writing into the original object.  This is by design, and should not be a problem EXCEPT for tests.  If a test compares the input to this function with the result, but the input was mutated in-situ, it may hide errors, resulting in tests that pass, but shouldn't.


/kind bug


```release-note
NONE
```

lavalamp commented:
All the tests modified in this PR appear to have been anticipating mutation in place. Are there other tests affected?

smarterclayton commented:
/hold

I remember designing this to allow registries to mutate, and this is a performance critical path.  We deserialize an object at the REST layer and pass it into storage to create.  Someone expecting create not to mutate is using it wrong.  I'd prefer not to change this path, but I agree that for test logic this should be easier to deal with.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -360,6 +360,9 @@ func finishNothing(context.Context, bool) {}
 // Create inserts a new item according to the unique key from the object.
 func (e *Store) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
+	// Make a copy so we never mutate the passed-in object (e.g. in tests)
+	obj = obj.DeepCopyObject()
```
Would be a regression in hot path serving performance - Create in apiserver should be called with net new objects anyway.  Is this ever a problem from outside of tests (someone calling storage create path in another storage call)?

thockin replied:
I don't think it's a REAL problem, just in tests.  Ergo my email to sig-api-machinery.

I don't think it's realistic to ask every test to pass a DeepCopy()

smarterclayton replied:
Why?  If you call create you need to understand what Create does.  You're giving it an object and get another object in return.  We could make it explicit that it's unsafe like Convert (UnsafeConvert mutates, Convert copies, everyone uses UnsafeConvert in hot path), which is our extant pattern for this concern.

Note that serialization mutates right now too...

thockin replied:
The more we disperse the responsibility, the more likely someone is to flub it.

IF we set a flag in `Store` that indicated "testing mode" then we could make this conditional.  That still depends on tests setting the flag, but that will get cargo-culted over time.  I dislike having test-only code in prod, but it's better than nothing?

BenTheElder replied:
FWIW we do have the `notest` build tag set on production binary builds so we could have this gated at compile time by that (ref:https://github.com/kubernetes/kubernetes/pull/92491)

I'm not sure that's the right answer, however.

thockin commented:
> All the tests modified in this PR appear to have been anticipating mutation in place. Are there other tests affected?

All the other tests in this file correctly look at the returned meta, rather than the original.  It's clear to me the test changes here are correct.  I don't buy that in-place mutation is semantically meaningful, though it may be important for perf (not measured).

Update already appears to do this DeepCopy, so I am skeptical of the perf assertion, too.

lavalamp commented:
> All the other tests in this file correctly look at the returned meta, rather than the original. It's clear to me the test changes here are correct. I don't buy that in-place mutation is semantically meaningful, though it may be important for perf (not measured).

If these three tests are the only ones affected, then I don't understand why you want to change this. They were verifying the correct things, admittedly in a slightly unpalatable way, but it still worked?

I thought you were going to show me tests that were `DeepEqual`ing original and created objects, and then I might have a different opinion. But given that this code is super low churn, and that there aren't actually bad tests, I'm currently not in favor of making this change.

I'm 100% in favor of making the Create documentation inform folks that it modifies in place.

thockin commented:
I have a WIP branch which beefs up service tests and removes the extra REST layer, which fails with this patch.  I'll add a commit here to show.

smarterclayton commented:
Services is definitely the last remaining holdout, but we haven't really hit this problem in the last 5 years in the stack (the mutate in place), so I'm kind of in the "I see no reason to change, but definitely reason to clarify, and if that clarification makes tests easier to write I'm happy".

thockin committed:
Service REST: Use DeepCopy() on Create() and fix tests

thockin commented:
If the core tests flubbed this, it seems like a trap to me.  I can't rationalize leaving a known trap like this.

lavalamp commented:
> If the core tests flubbed this, it seems like a trap to me. I can't rationalize leaving a known trap like this.

Where are the tests that flubbed this? I have not seen any yet?

thockin commented:
>> If the core tests flubbed this, it seems like a trap to me. I can't rationalize leaving a known trap like this.
>
> Where are the tests that flubbed this? I have not seen any yet?

3 flubs in https://github.com/kubernetes/kubernetes/pull/103545/files#diff-80a2fa04577a8670806fcb7ba3a96efe4ba8a8d8428123b83c64b0eb924bb894

I argue they are flubs because EVERY OTHER test in there was correct, except these 3.

lavalamp commented:
I do not agree those tests are wrong. They are checking an object which was modified for the modifications, in a way that would have caught actual problems.

thockin commented:
I'm very surprised.  What's the point of returning the created object if we are asserting we can modify it in place?

IMO we either document that Create() modifies the original passed-in object (thereby forcing every registry impl to take ownership of DeepCopying input) and "fix" ("break", IMO) all the other generic tests to just check the input object, or we make sure that input is not modified (in tests).

It seems very twilight zone to me to be arguing this - do you want broken tests?  This is how you get broken tests.

Lastly, perf:  I call hogwash.

Create calls `DeepCopy()` when calling `createValidation()` despite that being DOCUMENTED as "A ValidateObjectFunc may NOT transform the provided object".

It also calls `e.NewFunc()` in order to prep the return value, obviously copying again.

If we think perf matters here, then both of those are wrong.

lavalamp commented:
I agree that in theory, this should cause wrong tests. However this is low-churn code, and tomorrow is code freeze, and absent *actually broken* (not just ugly) tests, I'm not motivated to change it right now.

I am also skeptical that an additional deepcopy will cause a perf problem given that certain paths in the encoding/decoding stack do this already.

thockin commented:
I'm OK to sit on this until 1.23 cycle, if you have uncertainty about it.  IMO this falls under the "clearly not correct" heading and honestly tests which pass but shouldn't are a bigger risk than no tests at all.  Luckily the only such tests are within Service (but anyone could make this mistake).

smarterclayton commented:
> Create calls DeepCopy() when calling createValidation() despite that being DOCUMENTED as "A ValidateObjectFunc may NOT transform the provided object".

Agreed, but there's a fan out problem on validation (everyone who writes validation and gets it wrong) and a fan in problem on Create (only 1-2 places in codebase).

The deep copy there was justifiable paranoia - we kind of half agreed the better fix was a way to show that all validation functions were non mutating via fuzz but just never got that far.  That's still the right approach in my head.

thockin commented:
> a fan in problem on Create (only 1-2 places in codebase).

And a ton of tests which are all liable to get false-positives.  I am very open to better answers here :)

caesarxuchao commented:
/triage accepted

dims commented:
/milestone v1.23

per @thockin comment in https://github.com/kubernetes/kubernetes/pull/103545#issuecomment-875839134

thockin commented:
Now that 1.23 is in full swing - how do you all feel about this?

I posit 3 things:

1) k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go has some tests
that examine the input object as output, and some tests that examine the
returned object.  Ergo, it's clearly not universally understood that
Create() is allowed to modify the input object in-place.  So I think we
should AT LEAST fix those tests to be consistent.

2) There are "actually broken" tests in Service registry that this change
catches.  If this is the ONLY case (and it seems so) then I can
(begrudgingly) own this change inside service.  I know Service is a little
unusual in how it does its registry.  (I am trying to bring it more in-line
with the norm, which is how I hit this)

3) Perf-based arguments have no legs.  We wantonly copy objects "just in
case" and while it's disgusting (Go team - pretty please, const-pointer?)
it's understood and "safe" in the face of highly distributed API
ownership.  Yes *this* code is low-churn, but it's the 50+ registries that
are owned by other people that scare me.

So if the majority is still against this change (please indicate) I will
reduce it to just test "fixes" and comment updates about mutate-in-place,
and I will own it in the service registry.

WDYT?

On Thu, Jul 15, 2021 at 7:41 AM Davanum Srinivas ***@***.***>
wrote:

> /milestone v1.23
>
> per @thockin <https://github.com/thockin> comment in #103545 (comment)
> <https://github.com/kubernetes/kubernetes/pull/103545#issuecomment-875839134>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/103545#issuecomment-880751520>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVFXFCMAWPSTNJ7UZM3TX3XSPANCNFSM4766R5UQ>
> .
>

lavalamp commented:
If we *were* using C++ or Rust or something else with a larger type vocabulary, would we make this immutable/const? I don't think so, I think we would clearly pass ownership to the create function. My preference is still to make it very clear what is happening, e.g. rename the parameter "mutableObj" and add a note to the function description, and audit the tests so that they make a copy as needed.

This is 90% an aesthetic argument and 10% a performance argument -- yes we make deep copies a lot more than we should, which makes each additional one a smaller marginal cost. But in aggregate, we would generate way less garbage if we didn't make any copies. "We're already taking a -1% cost in 15 other places, so therefore the sixteeth place is fine also" is actually not a good argument if you consider the iterated version.

That said, I don't feel very strongly given that there's no imminent release.

thockin commented:
I feel like we WOULD use const on this, if the language supported it.

But now I am just arguing.  I'll let majority opinion of the subsystem
owners be the final word here.

On Fri, Aug 6, 2021, 11:29 AM Daniel Smith ***@***.***> wrote:

> If we *were* using C++ or Rust or something else with a larger type
> vocabulary, would we make this immutable/const? I don't think so, I think
> we would clearly pass ownership to the create function. My preference is
> still to make it very clear what is happening, e.g. rename the parameter
> "mutableObj" and add a note to the function description, and audit the
> tests so that they make a copy as needed.
>
> This is 90% an aesthetic argument and 10% a performance argument -- yes we
> make deep copies a lot more than we should, which makes each additional one
> a smaller marginal cost. But in aggregate, we would generate way less
> garbage if we didn't make any copies. "We're already taking a -1% cost in
> 15 other places, so therefore the sixteeth place is fine also" is actually
> not a good argument if you consider the iterated version.
>
> That said, I don't feel very strongly given that there's no imminent
> release.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/103545#issuecomment-894442506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVFMEQVZGRERS5DA72LT3QSZPANCNFSM4766R5UQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

lavalamp commented:
Like I say, I don't feel super strong. Happy to let @deads2k tiebreak.

thockin committed:
REST: Document mutable inputs on Create()

If one doesn't DeepCopy() on the way into Create, we can end up writing
into the original object.  This is by design, and should not be a
problem EXCEPT for tests.  If a test compares the input to this function
with the result, but the input was mutated in-situ, it may hide errors,
resulting in tests that pass, but shouldn't.

thockin committed:
Fix registry tests to look at result objects

thockin commented:
Does it make the proposal better or worse to only DeepCopy() in test mode? :)

lavalamp commented:
> Does it make the proposal better or worse to only DeepCopy() in test mode? :)

I feel like that could easily hide a system bug from the tests? So I would say worse.

thockin commented:
I'm going to interpret silence as a pocket veto.  I have pushed a different commit which merely comments on the risk (but still fixes the tests and fixes Service).  I will rename this PR.

thockin commented:
PTAL

lavalamp commented:
/lgtm
/approve

I personally would only take silence as an indication of business-- but anyway, I think it's fine to merge this, we can always make it stricter later if the documentation turns out not to help.

thockin commented:
I didn't mean that remark in anger - the sentiment isn't going my way and I am not willing to fight.  I can work around it locally.

lavalamp commented:
> I didn't mean that remark in anger

I didn't mean to imply that, sorry!

And I'll happily change my mind if we see problems in the future due to this.
